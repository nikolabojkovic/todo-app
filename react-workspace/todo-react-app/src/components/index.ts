export * from './TodoSettings';
export * from './Sort';
export * from './TodoList/TodoList';
export * from './Tabs/Tabs';
export * from './Paging/Paging';
export * from './AddTodo/AddTodo';
export * from './ConfirmModal/ConfirmModal';
export * from './Filter/FilterTodos';
export * from './ImportExport/ImportExport';
export * from './Pagination/PageSize';
export * from './Pagination/Pagination';
export * from './Search/Search';
export * from './TodoItem/TodoItem';
export * from './Loader/Loader';
export * from './TodoActions/TodoActions';
export * from './LoadMore/LoadMore';