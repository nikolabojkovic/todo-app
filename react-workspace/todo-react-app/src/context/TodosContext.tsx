import { createContext, useContext, useReducer } from "react";
import { IAction } from "../models/Action";
import { IPaging } from "../models/IPaging";
import { ITodoList } from "../models/ITodoList";
import { ITodo } from "../models/Todo";
 
export const TodosContext = createContext({} as ITodoList);
export const TodosDispatchContext = createContext(null as any);

type Props = {
  children: JSX.Element | JSX.Element[],
  todoList: ITodoList
}

export function TodoListProvider({ todoList, children }: Props) {
  const [state, dispatch] = useReducer(
    todoListReducer,
    todoList
  );

  return (
    <TodosContext.Provider value={state}>
      <TodosDispatchContext.Provider value={dispatch}>
        {children}
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>
  );
}

export function useTodoList() {
  return useContext(TodosContext);
}

export function useTodoListDispatch() {
  return useContext(TodosDispatchContext);
}

function todoListReducer(state: ITodoList, action: IAction) {
  switch (action.type) {
    case 'added': {
      const id = state.originalList.length >= 1 
      ? state.originalList
          .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[state.originalList.length - 1].id 
      : 0;
      const newTodo = {
        id: id + 1,
        title: action.payload.title,
        description: action.payload.description,
        completed: false, 
        createdAt: new Date()
      } as ITodo;

      return {
        ...state,
        originalList: [...state.originalList, newTodo],
        displayList: [...state.displayList, newTodo],
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount + 1,
          activePage: calculateActivePageOnAdd(state.paging),
          startIndex: (calculateActivePageOnAdd(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnAdd(state.paging) * state.paging.itemsPerPage
        } as IPaging
      } as ITodoList;
    }
    case 'changed': {
      return {
        ...state,
        originalList: state.originalList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        displayList: state.displayList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        paging: {...state.paging}
      } as ITodoList;
    }
    case 'deleted': {
      return {
        ...state,
        originalList: state.originalList.filter(t => t.id !== action.payload.id),
        displayList: state.displayList.filter(t => t.id !== action.payload.id),
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount - 1,
          activePage: calculateActivePageOnDelete(state.paging),
          startIndex: (calculateActivePageOnDelete(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnDelete(state.paging) * state.paging.itemsPerPage,  
        } as IPaging
      } as ITodoList;
    }
    case 'paging-updated': {
      return {
        ...state,
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as ITodoList
    }
    case 'searched': {
      return {
        ...state,
        displayList: [...action.payload.list],
        search: { searchTerm: action.payload.searchTerm },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      }
    }
    case 'searchTerm-updated': {
      return {
        ...state,
        search: { searchTerm: action.payload.searchTerm },
      }
    }
    case 'filtered': {
      return {
        ...state,
        displayList: [...action.payload.list],
        filter: {...action.payload.filter},
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      }
    }
    case 'sorted': {
      return {
        ...state,
        displayList: [...action.payload.list],
        sort: {...action.payload.sort},
        paging: {...state.paging} as IPaging
      }
    }
    case 'imported': {
      return {
        ...state,
        originalList: [...action.payload.list],
        displayList: [...action.payload.list],
        search: { searchTerm: '' },
        filter: {
          completed: false,
          uncompleted: false
         },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      }
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

function calculateActivePageOnDelete(paging: IPaging) {
  return Math.ceil((paging.totalCount - 1) / paging.itemsPerPage) < paging.activePage
    ? paging.activePage - 1 : paging.activePage;
}

function calculateActivePageOnAdd(paging: IPaging) {
  return Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) > paging.activePage
    ? Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) : paging.activePage;
}