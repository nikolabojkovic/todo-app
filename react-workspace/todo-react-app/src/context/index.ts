export * from './IState';
export * from './TodoListContext';
export * from './testData';