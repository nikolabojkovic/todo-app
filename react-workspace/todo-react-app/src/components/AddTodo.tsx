import { Button, Col, Container, Form, Row, Stack } from 'react-bootstrap';
import { useState } from 'react';
import { useTodoList, useTodoListDispatch } from '../context/TodosContext';
import { ITodo } from '../models/Todo';

export function AddTodo() {
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');

  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();
  const id = todoList.originalList.length >= 1 
    ? todoList.originalList
        .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[todoList.originalList.length - 1].id 
    : 0;

  return (
    <Form className="todo-background p-1">
      <Container fluid>
        <Row xs={1} sm={3}>
          <Col sm={5} className="p-2">
            <Form.Control 
              type="text" 
              placeholder="Enter title" 
              size="sm" 
              value={title}
              onChange={(e) => setTitle(e.target.value)} 
            />
          </Col>
          <Col sm={5} className="p-2">
            <Form.Control 
              type="text" 
              placeholder="Enter description" 
              size="sm" 
              value={description}
              onChange={(e) => setDescription(e.target.value)} 
            />
          </Col>
          <Col sm={2} className="p-2">
            <Button 
              variant="warning"
              style={{ 
                backgroundColor: '#FE9801',
                color: 'white', 
                minWidth: '90px',
                borderRadius: '20px',
                width: '100%'
              }} 
              size="sm"
              disabled={!title || title.trim() === '' || !description || description.trim() === ''}
              onClick={() => {
                dispatch({
                  type: 'added',
                  id: id + 1, 
                  title, 
                  description
                });
                setTitle('');
                setDescription('');
              }}
            >
              Add
            </Button>
          </Col>
        </Row>
      </Container>
    </Form>
  )
}