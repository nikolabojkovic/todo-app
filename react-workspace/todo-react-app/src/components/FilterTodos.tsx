import { Form, Stack } from "react-bootstrap";
import { useTodoList, useTodoListDispatch } from "../context/TodosContext";

export function FilterTodos() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function onFilter(completed: any, uncompleted: any) {
    dispatch({
      type: 'filtered',
      activePage: 1,
      filter: { 
        completed: completed,
        uncompleted: uncompleted
      }
    });
  }

  return (
    <Form className="App__filter d-flex justify-content-start align-items-center">
      <Stack direction="horizontal" gap={3}>       
        <Form.Check 
          type="switch"
          id={`completed`}
          label={`Completed`}
          checked={todoList.filter.completed}
          onChange={(e) => {
            onFilter(e.target.checked, todoList.filter.uncompleted);
          }}
        />   
        <Form.Check 
          type="switch"
          id={`uncompleted`}
          label={`Uncompleted`}
          checked={todoList.filter.uncompleted}
          onChange={(e) => {
            onFilter(todoList.filter.completed, e.target.checked);
          }}
        />         
      </Stack>
    </Form>
  )
}