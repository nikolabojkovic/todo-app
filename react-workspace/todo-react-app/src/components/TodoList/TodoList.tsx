import { useEffect } from 'react';
import { first } from 'rxjs';

import { ISort } from '../../models/ISort';
import { Loader } from '../Loader/Loader';
import { ITodo } from '../../models/Todo';
import { IAction, TodoActions } from '../../models/Action';

import { useTodoList, useTodoListDispatch } from '../../context/TodoListContext';

import { GetListProps, ITodoListProvider } from '../../providers/TodoProvider';

import { TodoItem } from '../TodoItem/TodoItem';
import { ListContainerType } from '../../models/ISettings';

type Props = {
  todoListProvider: ITodoListProvider
};

export function TodoList({ todoListProvider }: Props) {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  useEffect(() => {   
    dispatch({
      type: TodoActions.fetch,
      payload: {
        sort: {
          column: 'createdAt',
          direction: 'asc'
        } as ISort
      }
    } as IAction);
  }, [dispatch]);

  useEffect(() => {  
    if (todoList.effectTrigger
     && (todoList.effectTrigger.type === TodoActions.fetch
      || todoList.effectTrigger.type === TodoActions.filter
      || todoList.effectTrigger.type === TodoActions.search
      || todoList.effectTrigger.type === TodoActions.sort)) {
      dispatch({
        type: TodoActions.loadingStarted
      } as IAction);
      todoListProvider.getList({
        ...todoList.effectTrigger.payload,
      } as GetListProps)
      .pipe(first())
      .subscribe((list: ITodo[]) => {
        if (todoList.effectTrigger!.type === TodoActions.fetch) {
          dispatch({
            type: TodoActions.fetched,
            payload: {
              list: list
            }
          } as IAction);
        }

        if (todoList.effectTrigger!.type === TodoActions.filter) {
          dispatch({
            type: TodoActions.filtered,
            payload: {
              activePage: 1,
              list: list,
              filter: todoList.effectTrigger!.payload.filter
            }
          } as IAction);
        }

        if (todoList.effectTrigger!.type === TodoActions.search) {
          dispatch({
            type: TodoActions.searched,
            payload: {
              list: list,
              activePage: 1,
            }
          } as IAction);
        }

        if (todoList.effectTrigger!.type === TodoActions.sort) {
          dispatch({
            type: TodoActions.sorted,
            payload: {
              sort: todoList.effectTrigger!.payload.sort,
              list: list
            }
          } as IAction);
        }
      });
    }
  }, [todoList.effectTrigger, dispatch, todoListProvider]);

  useEffect(() => {
    if (todoList.effectTrigger
      && (todoList.effectTrigger.type === TodoActions.added
       || todoList.effectTrigger.type === TodoActions.changed
       || todoList.effectTrigger.type === TodoActions.deleted
       || todoList.effectTrigger.type === TodoActions.imported)) {
      todoListProvider.saveList(todoList.originalList).pipe(first()).subscribe();
     }
  }, [todoList.effectTrigger, todoList.originalList, todoListProvider]);

  function getDisplayList() {
    if (todoList.settings.general.isPaginationEnabled) {
      return todoList.displayList
        .slice(todoList.paging.startIndex, todoList.paging.endIndex)
        .map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />);
    }
      
    return todoList.displayList.map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />);
  }

  return (
    <>
    { todoList.activeTab !== 'settings' && 
      <main className="App__todo-list">
        { 
          todoList.isLoading 
          ? <Loader />
          : <section 
              className='' 
              style={ 
                todoList.settings.general.listSizeType === ListContainerType.Fixed 
                ? { 
                    height: todoList.settings.general.fixedListSize,
                    overflow: 'overlay'
                  } 
                : {}
                }>
              {
                (todoList.paging.totalCount > 0 ?
                  getDisplayList() 
                : <div className='text-light mt-5 mb-5 fade-in'>No data</div>)
              }
            </section>
        }
      </main>
    }
    </>
  );
}