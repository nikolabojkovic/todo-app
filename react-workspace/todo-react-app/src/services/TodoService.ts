import { inMemoryTodoListTestData } from "../context/testData";
import { IFilter } from "../models/IFilter";
import { ISort } from "../models/ISort";
import { TodoList } from "../models/ITodoList";
import { ITodo } from "../models/Todo";

export class TodoService {

  private todoListName: string = 'todo-list';

  getTodoList() {
    if (localStorage.getItem(this.todoListName) === undefined 
     || localStorage.getItem(this.todoListName) === null) {
      return inMemoryTodoListTestData;
    }
    
    const todos = this.sort(JSON.parse(localStorage.getItem(this.todoListName) ?? "[]") as ITodo[], {
      column: 'createdAt',
      direction: 'asc'
    } as ISort);
  
    let todoList = new TodoList(todos);

    return todoList;
  }

  saveTodoList(list: ITodo[]): void {
    localStorage.setItem(this.todoListName, JSON.stringify(list));
  }
  
  search(list: ITodo[], searchTerm: string,): ITodo[] {
    let filteredList = list;
  
    if (searchTerm !== '') {
      filteredList = list.filter((todo: ITodo) => 
        todo.title.trim()
                  .toLocaleLowerCase()
                  .includes(searchTerm.trim()
                                      .toLocaleLowerCase()) 
     || todo.description.trim()
                        .toLocaleLowerCase()
                        .includes(searchTerm.trim()
                                            .toLocaleLowerCase()));
    }
  
    return filteredList;
  }
  
  filter(list: ITodo[], filter: IFilter) : ITodo[] {
    let filteredList = list;
  
    if (filter && filter.completed && filter.uncompleted) {
      return [...filteredList];
    }
  
    if (filter?.completed) {
      filteredList = filteredList.filter((todo: ITodo) => todo.completed === true);
    }
  
    if (filter?.uncompleted) {
      filteredList = filteredList.filter((todo: ITodo) => todo.completed === false);
    }
  
    return [...filteredList];
  }
  
  sort(list: ITodo[], sort: ISort) : ITodo[] {
    let sortResult = [];
    
    if (sort.column === 'createdAt') {
      if (sort.direction === 'asc') {
        sortResult = [...list.sort((a: ITodo, b: ITodo) => a.createdAt > b.createdAt ? 1 : -1)]
      } else {
        sortResult = [...list.sort((a: ITodo, b: ITodo) => a.createdAt < b.createdAt ? 1 : -1)]
      }
  
      return sortResult;
    }
  
    if (sort.direction === 'asc') {
      sortResult = [...list.sort((a: any, b: any) => a[sort.column] > b[sort.column] ? 1 : -1)]
    } else {
      sortResult = [...list.sort((a: any, b: any) => a[sort.column] < b[sort.column] ? 1 : -1)]
    }
  
    return sortResult;
  }
}

export const todoServiceInstance = new TodoService();