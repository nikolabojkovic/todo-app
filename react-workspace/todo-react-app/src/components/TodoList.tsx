import { useEffect } from 'react'
import { AddTodo } from './AddTodo';
import { TodoItem } from './TodoItem';
import { Paging } from './Paging';
import { Search } from './Search';
import { useTodoList } from '../context/TodosContext';
import { Tabs } from './Tabs';
import { FilterTodos } from './FilterTodos';
import { Sorting } from './Sorting';
import { ImportExport } from './ImportExport';
import { ITodo } from '../models/Todo';

export function TodoList() {
  const todoList = useTodoList();

  // TODO: move save to context
  useEffect(() => {
    localStorage.setItem('todo-list', JSON.stringify(todoList.originalList));
  }, [todoList.originalList]);

  return (
    <main className="App__todo-list">
      <Tabs>
        <AddTodo key="add-todo" />
        <Search 
          key="search-todos" 
          placeholder='Search by title or description' 
        />
        <FilterTodos key="filter-todos" />
        <ImportExport key="import-export" />
      </Tabs>
      <Sorting/>
      {todoList.displayList
        .slice(todoList.paging.startIndex, todoList.paging.endIndex)
        .map((todo: ITodo) => <TodoItem key={todo.id} todo={todo} />)}
      <Paging/>
    </main>
  );
}