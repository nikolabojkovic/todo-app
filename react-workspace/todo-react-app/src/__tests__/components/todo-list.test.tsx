import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { TodoList } from '../../components/TodoList';
import { TodoListProvider } from '../../context/TodosContext';
import { TodoService } from '../../services/TodoService';

const todoService = new TodoService();

describe('todo list rendered', () => {
  it('match snapshot', () => {
    const todoList = todoService.getTodoList();
    const todoListJsxElements = (<TodoListProvider todoList={todoList}>
                                   <TodoList/>
                                 </TodoListProvider>);
    const tree = renderer.create(
      todoListJsxElements
    ).toJSON();
    render(todoListJsxElements);
    expect(tree).toMatchSnapshot();
  });
});
