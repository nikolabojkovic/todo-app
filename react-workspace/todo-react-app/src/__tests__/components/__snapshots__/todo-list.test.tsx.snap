// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo list rendered match snapshot 1`] = `
<main
  className="App__todo-list"
>
  <div
    className="App__todo-list__item"
  >
    <div
      className="hstack gap-3"
    >
      <div>
        <div
          className="App__todo-list__item-title"
        >
          Task 1
        </div>
        <div
          className="App__todo-list__item-description"
        >
          <span>
            Description 1
          </span>
          <span>
             Fri Feb 04 2022 
          </span>
        </div>
      </div>
      <button
        className="ms-auto confirm-button btn btn-outline-success btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Complete
      </button>
      <button
        className="confirm-button btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="App__todo-list__item"
  >
    <div
      className="hstack gap-3"
    >
      <div>
        <div
          className="App__todo-list__item-title--completed"
        >
          Task 2
        </div>
        <div
          className="App__todo-list__item-description--compleated"
        >
          <span>
            Description 2
          </span>
          <span>
             Sat Feb 05 2022 
          </span>
        </div>
      </div>
      <button
        className="ms-auto confirm-button btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="App__todo-list__item"
  >
    <div
      className="hstack gap-3"
    >
      <div>
        <div
          className="App__todo-list__item-title"
        >
          Task 3
        </div>
        <div
          className="App__todo-list__item-description"
        >
          <span>
            Description 3
          </span>
          <span>
             Sun Feb 06 2022 
          </span>
        </div>
      </div>
      <button
        className="ms-auto confirm-button btn btn-outline-success btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Complete
      </button>
      <button
        className="confirm-button btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="App__todo-list__item"
  >
    <div
      className="hstack gap-3"
    >
      <div>
        <div
          className="App__todo-list__item-title"
        >
          Task 4
        </div>
        <div
          className="App__todo-list__item-description"
        >
          <span>
            Description 4
          </span>
          <span>
             Mon Feb 07 2022 
          </span>
        </div>
      </div>
      <button
        className="ms-auto confirm-button btn btn-outline-success btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Complete
      </button>
      <button
        className="confirm-button btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    className="App__todo-list__item"
  >
    <div
      className="hstack gap-3"
    >
      <div>
        <div
          className="App__todo-list__item-title"
        >
          Task 5
        </div>
        <div
          className="App__todo-list__item-description"
        >
          <span>
            Description 5
          </span>
          <span>
             Fri Apr 08 2022 
          </span>
        </div>
      </div>
      <button
        className="ms-auto confirm-button btn btn-outline-success btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Complete
      </button>
      <button
        className="confirm-button btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
  <section
    className="paging-container p-0 mt-2"
  >
    <div
      className="container-fluid"
    >
      <div
        className="row row-cols-sm-3 row-cols-1"
      >
        <div
          className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center col-sm-3"
        >
          Page 
          1
           of 
          2
        </div>
        <div
          className="p-3 pt-2 pb-2 col-sm-4"
        >
          <div
            className="justify-content-center hstack gap-2"
          >
            <div
              className="ms-sm-auto"
            >
              Page size:
               
            </div>
            <div
              className=""
            >
              <select
                aria-label="Page size"
                className="form-select form-select-sm"
                id="page-size"
                name="page-size"
                onChange={[Function]}
              >
                <option
                  value={5}
                >
                  5
                </option>
                <option
                  value={10}
                >
                  10
                </option>
                <option
                  value={50}
                >
                  50
                </option>
                <option
                  value={100}
                >
                  100
                </option>
              </select>
            </div>
          </div>
        </div>
        <div
          className="p-3 pt-2 pb-2 d-flex justify-content-sm-end justify-content-center col-sm-5"
        >
          <ul
            className="pagination pagination-sm"
          >
            <li
              className="page-item disabled"
            >
              <span
                className="page-link"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  «
                </span>
                <span
                  className="visually-hidden"
                >
                  First
                </span>
              </span>
            </li>
            <li
              className="page-item disabled"
            >
              <span
                className="page-link"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  ‹
                </span>
                <span
                  className="visually-hidden"
                >
                  Previous
                </span>
              </span>
            </li>
            <li
              className="page-item active"
            >
              <span
                className="page-link"
                onClick={[Function]}
              >
                1
                <span
                  className="visually-hidden"
                >
                  (current)
                </span>
              </span>
            </li>
            <li
              className="page-item"
            >
              <a
                className="page-link"
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                2
              </a>
            </li>
            <li
              className="page-item"
            >
              <a
                className="page-link"
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  aria-hidden="true"
                >
                  ›
                </span>
                <span
                  className="visually-hidden"
                >
                  Next
                </span>
              </a>
            </li>
            <li
              className="page-item"
            >
              <a
                className="page-link"
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                <span
                  aria-hidden="true"
                >
                  »
                </span>
                <span
                  className="visually-hidden"
                >
                  Last
                </span>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </section>
</main>
`;
